Percolation.java

```diff
@@ -1,14 +1,135 @@
 import edu.princeton.cs.algs4.StdRandom;
-import edu.princeton.cs.algs4.StdStats;
+import  edu.princeton.cs.algs4.StdStats;
 import edu.princeton.cs.algs4.WeightedQuickUnionUF;
 
 public class Percolation {
-   public Percolation(int n)                // create n-by-n grid, with all sites blocked
-   public    void open(int row, int col)    // open site (row, col) if it is not open already
-   public boolean isOpen(int row, int col)  // is site (row, col) open?
-   public boolean isFull(int row, int col)  // is site (row, col) full?
-   public     int numberOfOpenSites()       // number of open sites
-   public boolean percolates()              // does the system percolate?
-
-   public static void main(String[] args)   // test client (optional)
-}
\ No newline at end of file
+
+  // UF used for percolates test
+  private WeightedQuickUnionUF PercolateTest;   
+
+  // UF used for connection site test
+  // open site in the top row
+  private WeightedQuickUnionUF TestFull;
+
+  // siteStatus stores every site's open or block
+  // true means open,false means block
+  private boolean[] siteStatus;
+
+  // percolation experiment's grid width and height
+  private int SIZE;
+
+  // 1D of site connected to all sites in first row
+  private int H20_SURFACE;
+
+  // 1D of site connected to all sites in last row
+  private int OIL_BELOW;
+
+  // create N-by-N grid, with all sites blocked
+  public Percolation(int N) {
+    SIZE = N;
+
+    // the last two site for H20_SURFACE and OIL_BELOW
+    PercolateTest = new WeightedQuickUnionUF(SIZE * SIZE + 2);
+
+    // the last site is for H20_SURFACE
+    TestFull = new WeightedQuickUnionUF(SIZE * SIZE + 1);
+
+    H20_SURFACE = SIZE * SIZE;
+    OIL_BELOW = SIZE * SIZE + 1;
+
+    // initialize all site status blocked
+    siteStatus = new boolean[SIZE*SIZE + 2];
+    for (int i = 0; i < SIZE*SIZE; i++) {
+      siteStatus[i] = false;
+    }
+
+    // initialize the two sites open
+    siteStatus[H20_SURFACE] = true;
+    siteStatus[OIL_BELOW] = true;
+
+    // for the UF for percolate test and the UF for test full
+    // connect each site in first row to H20_SURFACE
+    for (int i = 0; i < SIZE; i++) {
+      PercolateTest.union(H20_SURFACE, i);
+      TestFull.union(H20_SURFACE, i);
+    }
+
+    // only for the UF used for percolate test
+    // connect each site in last row to OIL_BELOW
+    for (int i = (SIZE-1)*SIZE; i < SIZE*SIZE; i++) {
+      PercolateTest.union(OIL_BELOW, i);
+    }
+
+  }
+
+  // open site (row i, column j) if it is not already
+  // i,j between 1 and N
+  public void open(int i, int j) {
+    if (!validIndex(i, j)) {
+      throw new IndexOutOfBoundsException("row or/and column index out of bounds");
+    }
```

Comment:
You're repeating this pattern three times, you could extract it to a checkIndex method and just call that one...